File: NthNode.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        fast = slow = dummy

        #Move fast pointer to next step
        for i in range(n + 1):
            fast = fast.next

        #Move both pointers until fast reaches the end
        while fast:
            fast = fast.next
            slow = slow.next

        #Remove the target node
        slow.next = slow.next.next

        return dummy.next

    #Helper function to print thre linked list
    def printList(head: Optional[ListNode]):
        while head:
            print(head.val, end="->")
            head = head.next
        print("None")
        

File: NthNode.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr)
    {

    }
};

class Solution {
    public:
        ListNode* removeNthFromEnd(ListNode* head, int n)
        {
            ListNode* dummy = new ListNode(0);
            dummy->next = head;
            ListNode* fast = dummy;
            ListNode* slow = dummy
        }
}